src/fd.cpp

```diff
@@ -0,0 +1,248 @@
+#ifdef _WIN32
+#ifndef FD_SETSIZE
+#define FD_SETSIZE 2048  // Set max fds - only for select() fallback
+#endif
+#include <winsock2.h>
+#else
+#include <poll.h>
+#endif
+#include <Rcpp.h>
+#include <unistd.h>
+#include <cstdlib>
+#include <atomic>
+#include "tinycthread.h"
+#include "later.h"
+#include "callback_registry_table.h"
+
+#if R_VERSION < R_Version(4, 5, 0)
+inline SEXP R_mkClosure(SEXP formals, SEXP body, SEXP env) {
+  SEXP fun = Rf_allocSExp(CLOSXP);
+  SET_FORMALS(fun, formals);
+  SET_BODY(fun, body);
+  SET_CLOENV(fun, env);
+  return fun;
+}
+#endif
+
+extern CallbackRegistryTable callbackRegistryTable;
+extern SEXP later_fdcancel;
+extern SEXP later_invisibleSymbol;
+
+#define LATER_INTERVAL 1024
+
+#ifdef _WIN32
+#define POLL_FUNC WSAPoll
+#else
+#define POLL_FUNC poll
+#endif
+
+typedef struct ThreadArgs_s {
+  std::shared_ptr<std::atomic<bool>> flag;
+  std::unique_ptr<std::vector<int>> fds;
+  SEXP callback;
+  double timeout;
+  int rfds, wfds, efds;
+  int num_fds;
+  int loop;
+} ThreadArgs;
+
+static void later_callback(void *arg) {
+
+  std::unique_ptr<std::shared_ptr<ThreadArgs>> argsptr(static_cast<std::shared_ptr<ThreadArgs>*>(arg));
+  std::shared_ptr<ThreadArgs> args = *argsptr;
+  const bool flag = args->flag->load();
+  args->flag->store(true);
+  if (!flag) {
+    std::memcpy((void *) DATAPTR_RO(CADR(args->callback)), args->fds->data(), args->num_fds * sizeof(int));
+    Rf_eval(args->callback, R_GlobalEnv);
+  }
+  R_ReleaseObject(args->callback);
+
+}
+
+// CONSIDER: if necessary to add method for HANDLES on Windows. Would be different code to SOCKETs.
+// TODO: implement re-usable background thread.
+static int wait_thread(void *arg) {
```

Comment:
I find this (pretty important) function difficult to understand primarily because of the length, I think. Can I suggest that it be broken up a little, especially that the two select/poll implementations be a separate function with as simple arguments as possible?

Or how about if this code is just written for `poll()`, and then include a poll polyfill? Here, I had Claude 3.5 Sonnet write this for you :smile:

poll_polyfill.h
```cpp
#ifndef POLL_POLYFILL_H
#define POLL_POLYFILL_H

#ifdef __cplusplus
extern "C" {
#endif

/* Check if we actually need the polyfill */
#if defined(_WIN32) && _WIN32_WINNT < 0x0600 /* Pre-Vista Windows */ || \
    (defined(__unix__) && !defined(HAVE_POLL))

#include <stddef.h>

#ifdef _WIN32
    typedef unsigned long nfds_t;
#else
    #include <sys/types.h>
    typedef unsigned long nfds_t;
#endif

/* Standard poll flags if not already defined */
#ifndef POLLIN
    #define POLLIN      0x0001    /* There is data to read */
    #define POLLPRI     0x0002    /* There is urgent data to read */
    #define POLLOUT     0x0004    /* Writing now will not block */
    #define POLLERR     0x0008    /* Error condition */
    #define POLLHUP     0x0010    /* Hung up */
    #define POLLNVAL    0x0020    /* Invalid request: fd not open */
    #define POLLRDNORM  0x0040    /* Normal data may be read */
    #define POLLRDBAND  0x0080    /* Priority data may be read */
    #define POLLWRNORM  0x0100    /* Writing now will not block */
    #define POLLWRBAND  0x0200    /* Priority data may be written */
#endif

/* Standard pollfd structure if not already defined */
#ifndef POLLFD_DEFINED
#define POLLFD_DEFINED
struct pollfd {
    int   fd;      /* file descriptor */
    short events;  /* requested events */
    short revents; /* returned events */
};
#endif

/* The actual poll function */
int poll(struct pollfd *fds, nfds_t nfds, int timeout);

#endif /* Poll needed check */

#ifdef __cplusplus
}
#endif

#endif /* POLL_POLYFILL_H */
```

poll_polyfill.cpp
```cpp
#include "poll_polyfill.h"

/* Check if we actually need the polyfill */
#if defined(_WIN32) && _WIN32_WINNT < 0x0600 /* Pre-Vista Windows */ || \
    (defined(__unix__) && !defined(HAVE_POLL))

#include <errno.h>
#include <string.h>
#include <limits.h>

#ifdef _WIN32
    #include <winsock2.h>
    #pragma comment(lib, "ws2_32.lib")
#else
    #include <sys/select.h>
    #include <sys/time.h>
    #include <sys/types.h>
    #include <unistd.h>
    #include <fcntl.h>
#endif

namespace {
    // Helper class for automatic initialization of Winsock on Windows
    #ifdef _WIN32
    class WinsockInitializer {
    public:
        WinsockInitializer() {
            WSADATA wsaData;
            WSAStartup(MAKEWORD(2, 2), &wsaData);
        }
        ~WinsockInitializer() {
            WSACleanup();
        }
    };
    static WinsockInitializer winsockInit;
    #endif

    // Helper function to check if a file descriptor is valid
    bool is_fd_valid(int fd) {
        #ifdef _WIN32
            int sock_err = 0;
            int sock_err_len = sizeof(sock_err);
            return getsockopt(fd, SOL_SOCKET, SO_ERROR, 
                          reinterpret_cast<char*>(&sock_err), 
                          &sock_err_len) != SOCKET_ERROR;
        #else
            return fcntl(fd, F_GETFL) != -1 || errno != EBADF;
        #endif
    }

    // Helper function to handle platform-specific sleep
    void platform_sleep(int timeout_ms) {
        #ifdef _WIN32
            Sleep(timeout_ms);
        #else
            usleep(timeout_ms * 1000);
        #endif
    }
}

extern "C" int poll(struct pollfd *fds, nfds_t nfds, int timeout) {
    // Validate input parameters
    if (fds == nullptr && nfds > 0) {
        errno = EFAULT;
        return -1;
    }

    // Check against FD_SETSIZE limit
    if (nfds > FD_SETSIZE) {
        errno = EINVAL;
        return -1;
    }

    // Initialize fd sets
    fd_set readfds, writefds, exceptfds;
    FD_ZERO(&readfds);
    FD_ZERO(&writefds);
    FD_ZERO(&exceptfds);

    // Find highest fd and set up fd sets
    int max_fd = -1;
    bool has_valid_fd = false;

    for (nfds_t i = 0; i < nfds; ++i) {
        // Clear return events
        fds[i].revents = 0;

        if (fds[i].fd < 0) {
            continue;
        }

        // Check for invalid descriptors early
        if (!is_fd_valid(fds[i].fd)) {
            fds[i].revents |= POLLNVAL;
            continue;
        }

        has_valid_fd = true;
        max_fd = (fds[i].fd > max_fd) ? fds[i].fd : max_fd;

        // Set up fd sets based on requested events
        if (fds[i].events & (POLLIN | POLLRDNORM | POLLRDBAND | POLLPRI)) {
            FD_SET(fds[i].fd, &readfds);
        }
        if (fds[i].events & (POLLOUT | POLLWRNORM | POLLWRBAND)) {
            FD_SET(fds[i].fd, &writefds);
        }
        FD_SET(fds[i].fd, &exceptfds);  // Always check for errors
    }

    // Handle case with no valid file descriptors
    if (!has_valid_fd) {
        if (timeout >= 0) {
            platform_sleep(timeout);
            return 0;
        }
        errno = EINVAL;
        return -1;
    }

    // Set up timeout structure
    struct timeval tv, *tvp = nullptr;
    if (timeout >= 0) {
        tv.tv_sec = timeout / 1000;
        tv.tv_usec = (timeout % 1000) * 1000;
        tvp = &tv;
    }

    // Perform select() operation
    int ready = select(max_fd + 1, &readfds, &writefds, &exceptfds, tvp);
    
    if (ready == -1) {
        return -1;  // Error is already in errno
    }
    if (ready == 0) {
        return 0;   // Timeout
    }

    // Process results
    int total_events = 0;
    for (nfds_t i = 0; i < nfds; ++i) {
        if (fds[i].fd < 0) {
            continue;
        }

        // Skip if already marked invalid
        if (fds[i].revents & POLLNVAL) {
            ++total_events;
            continue;
        }

        // Map select() results to poll() events
        if (FD_ISSET(fds[i].fd, &readfds)) {
            fds[i].revents |= (fds[i].events & (POLLIN | POLLRDNORM));
        }
        if (FD_ISSET(fds[i].fd, &writefds)) {
            fds[i].revents |= (fds[i].events & (POLLOUT | POLLWRNORM));
        }
        if (FD_ISSET(fds[i].fd, &exceptfds)) {
            fds[i].revents |= POLLERR;
            // Check for OOB data
            if (fds[i].events & POLLPRI) {
                fds[i].revents |= POLLPRI;
            }
        }

        // Recheck descriptor validity
        if (!is_fd_valid(fds[i].fd)) {
            fds[i].revents |= POLLNVAL;
        }

        if (fds[i].revents) {
            ++total_events;
        }
    }

    return total_events;
}

#endif /* Poll needed check */

Reply:
If you don't like this idea (polyfill), I won't argue. It's what I would do but I would understand if you don't like the tradeoff.

Reply:
Thanks for the review and comments Joe! I've split the function into two versions for now as you suggested in [9ceaf14](https://github.com/r-lib/later/pull/190/commits/9ceaf141556f483259763b0e7387c2eb4cc80d63). Over time, this will sort itself out when we don't need to support R <= 4.1.

As for the polyfill... I got excited when I first saw your comment, thought it was some way to bridge to a poll() implementation that it could find...

I'm still looking at this and the couple of remaining points!

Reply:
> If you don't like this idea (polyfill), I won't argue. It's what I would do but I would understand if you don't like the tradeoff.

I must say that polyfill didn't initially appeal (because of the inefficiency), but the idea is growing on me as it's not _that_ much work to map from a `pollfd()` to an `fd_set()`.

Reply:
And I now have an amazing 3 line 'polyfill' that solves this in [c46399c](https://github.com/r-lib/later/pull/190/commits/c46399caafe42d8e2c9e40f34ba62f9084b6e5f0) 😄😄😄.

Reply:
That makes so much sense!